Add full Go 1.21 support

eskultety opened the pull request:
# Maintainers will complete the following section

This PR enables full Go 1.21 support by the means of instructing Go to download the desired toolchain via`GOTOOLCHAIN=auto`. This is limited to 1.21 only, so 1.22 will be explicitly refused (1.22 will be added by a standalone PR).

Notes:
- first 3 patches are some generic tiny fixes
- `EnvironmentVariable` was adjusted to act as a more generic resolvable template (e.g. to allow `GOPROXY=file://${GOMODCACHE}/foo`)
- Go 1.21 full support is enabled using `GOTOOLCHAIN` and `GOPROXY`

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

chmeliik commented on cachi2/core/models/output.py:
```diff
@@ -48,7 +49,29 @@ def resolve_value(self, mappings: Dict[str, str]) -> str:
+        for i, m in enumerate(mappings):
+            log.debug(f"EnvironmentVariable resolution iteration {i}.: {ret}")
+
+            placeholders_seen.append(m)
```
The nested variable resolution is very cool :+1: 

But in the current state, cycle detection doesn't work correctly (depends on the order of the `mapping`)
```python
>>> e = EnvironmentVariable(name="A", value="$C")
>>> e.resolve_value({"C": "$B", "B": "1"})
'1'
>>> e.resolve_value({"B": "1", "C": "$B"})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/acmiel/RedHat/Cachito/cachi2/cachi2/core/models/output.py", line 66, in resolve_value
    raise Cachi2Error(
cachi2.core.errors.Cachi2Error: Detected a cycle in environment variable expansion of 'A'
```

eskultety replied:
Doh! Nasty bug, shameful...@chmeliik thanks for pointing this out!

brunoapimentel replied:
I know you'll try to solve this meticulously first thing in the morning, but I wonder if we need nested variable resolution in the first place. I does look nice, but it adds quite some complexity, and maybe we don't need it right now.

Our current use case can be solved if we declare GOPROXY as:
```
file://${output_dir}/deps/gomod/pkg/mod/cache/download
```
... and we can hold on adding nested variable resolution for when a use case that can't be solved this way comes up.

eskultety replied:
@chmeliik updated with IMO a more robust solution, see if you can spot another resolution problem. I also tweaked and rearranged the test data so that your finding is covered.

eskultety replied:
@brunoapimentel I understand the motivation and what you suggest was my initial approach, but then I figured that I hate maintaining duplicate static strings which leads to the need to fix them eventually at all places. It all comes down to immediate effect vs hardening. IMO it is the immediate impact fixes that usually lead to and add up to the technical debt that will likely never be addressed, but there surely is a line to be drawn. If you feel strongly about the added complexity (which won't even be exercised 99% of the cases) being too high, then I'll drop the change no problem.

brunoapimentel replied:
I don't feel strongly about it, my reasoning was mostly to avoid extra work, since we will need to reach a reasonable solution with this nested variable resolution algorithm.

chmeliik commented on cachi2/core/models/output.py:
```diff
@@ -43,12 +44,53 @@ def resolve_value(self, mappings: Dict[str, str]) -> str:
+                        "variables altogether."
+                    ),
+                )
+            substituted |= get_placeholders(t_old) - p_new
```
If I understand correctly, this is needed for cases such as:

```python
before_substitution = "${known_variable} and ${unknown_variable}"
after_substitution = "some_value and ${unknown_variable}"
```
We don't want the `unknown_variable` to cause false-positive cycle detection errors, so we avoid marking it as substituted

I wonder if this can lead to false negatives as well though
```python
before_substitution = "${foo} and ${bar}"
after_substitution = "${bar} and ${foo}"
# get_placeholders(before_substitution) - get_placeholders(after_substitution) == empty set ?
```

eskultety replied:
> I wonder if this can lead to false negatives as well though
>```
>before_substitution = "${foo} and ${bar}"
>after_substitution = "${bar} and ${foo}">
># get_placeholders(before_substitution) - get_placeholders(after_substitution) == empty set
>```

@chmeliik if we add an additional `before_substitution != after_substitution` that would take care of the problem wouldn't it?

chmeliik replied:
Hmm, not sure. The cycles can be more complicated
```python
mappings = {
    "cycle_part1": "$cycle_part3",
    "cycle_part2": "$cycle_part1",
    "cycle_part3": "$cycle_part2"
}
```
```python
"$cycle_part1 $cycle_part2"
# ->
"$cycle_part3 $cycle_part1"  # substituted = {"cycle_part2"}
# ->
"$cycle_part2 $cycle_part3"  # substituted = {"cycle_part2", "cycle_part1"}
# cycle detected
```
In this case I believe the cycle will be caught after the second iteration (even though it could be caught after the first one), but I'm not sure if every cycle will be caught

chmeliik replied:
Maybe, rather than subtracting the new placeholders from the `substituted` set, a more robust solution would be to intersect `substituted` with the known `mappings`?

```python
substituted |= get_placeholders(t_old) & mappings.keys()
```

That should still cover the "don't mark unknown variables as substituted" case and shouldn't interfere with cycle detection for known variables

eskultety replied:
> Hmm, not sure. The cycles can be more complicated
> 
> ```python
> mappings = {
>     "cycle_part1": "$cycle_part3",
>     "cycle_part2": "$cycle_part1",
>     "cycle_part3": "$cycle_part2"
> }
> ```
> 
> ```python
> "$cycle_part1 $cycle_part2"
> # ->
> "$cycle_part3 $cycle_part1"  # substituted = {"cycle_part2"}
> # ->
> "$cycle_part2 $cycle_part3"  # substituted = {"cycle_part2", "cycle_part1"}
> # cycle detected
> ```
> 
> In this case I believe the cycle will be caught after the second iteration (even though it could be caught after the first one), but I'm not sure if every cycle will be caught

I guess there is an algorithm that could catch it after the first iteration, but as long as it's detected it's not an issue.

eskultety replied:
> Maybe, rather than subtracting the new placeholders from the `substituted` set, a more robust solution would be to intersect `substituted` with the known `mappings`?
> 
> ```python
> substituted |= get_placeholders(t_old) & mappings.keys()
> ```
> 
> That should still cover the "don't mark unknown variables as substituted" case and shouldn't interfere with cycle detection for known variables

It felt a bit counter intuitive, but yeah, this seems to work, maybe I just rename it to `seen` rather than substituted that actually might describe it better.

chmeliik replied:
IMO `substituted` is still more accurate. In each iteration, every placeholder in the template string really does get substituted - unless it has no mapping in `mappings`

eskultety commented:
since v2:
- changed detection of variable resolution cycles per the review comments
- forced setting `GOTOOLCHAIN=local` for cases where `go base version > go.mod version > 1.21` (@taylormadore this relates to https://github.com/containerbuildsystem/cachi2/issues/501#issuecomment-2008970964, do you think it will be enough for now?)

taylormadore commented on cachi2/core/package_managers/gomod.py:
```diff
@@ -718,6 +719,57 @@ def _find_missing_gomod_files(source_path: RootedPath, subpaths: list[str]) -> l
+            go = Go(release="go1.21.0")
+        elif go_base_version > go_mod_version:
+            # we have a newer toolchain than the requirement, so we don't want GOTOOLCHAIN=auto
+            env["GOTOOLCHAIN"] = "local"
```
Is this set just for the prefetch? I think we'll need to provide it in the request output

eskultety replied:
darn, yep, good eyes!

eskultety replied:
In fact, this shouldn't even happen here, but more changes will be needed for `_fetch_gomod_source` to be able to make this kind of decision.

eskultety replied:
Based on the discussion in https://github.com/containerbuildsystem/cachi2/issues/501, let's keep `GOTOOLCHAIN=auto` only for prefetch.

eskultety commented:
since v3:

- `GOTOOLCHAIN=auto` is only used for dependency fetching, but it's not enforced upon users by the means of `generate-env`
- `toolchain` keyword in `go.mod` file is considered as well when checking the maximum supported version violation (due to the use of GOTOOLCHAIN=auto)
- container image is preinstalled with a static 1.21.0 Go SDK along with 1.20
- some minor debug log adjustments
- rebase on main

eskultety commented:
since v4:
- add documentation for Go 1.21 full support

eskultety commented:
@brunoapimentel @taylormadore sorry for the delay, please have another look at the changes. I'll work on some integration tests in the meantime.

eskultety commented:
Since v5:
- fix integration tests by generating fresh mock data

brunoapimentel commented on Containerfile:
```diff
@@ -34,10 +34,12 @@ RUN npm install && \
-    mkdir /usr/local/go && \
-    mv "$HOME/sdk/go1.20" /usr/local/go && \
-    rm -rf "$HOME/go" "$HOME/.cache/go-build/"
+RUN for go_ver in "go1.20" "go1.21.0"; do \
```
Not a problem for this PR, but I wonder how we'll enable hermetic Cachi2 builds of Cachi2 since we need to pull these toolchains from the internet. My first idea was to try to create some sort of `go-deps` directory, but then, how can we ask Go to fetch toolchains <= 1.21.0? For toolchains > 1.21.0, the download is easy to setup. 

In any case, we don't need to solve this now, I'm just thinking out loud.

eskultety replied:
> how can we ask Go to fetch toolchains <= 1.21.0? For toolchains > 1.21.0, the download is easy to setup.

Last time I looked at the cache, it seemed like we'd have to reverse engineer how Go creates the directory structure in GOMODCACHE and copy files over there by ourselves from the SDKs downloaded the same way as in this code block. I think a separate discussion is in place and things should become a bit easier if we conclude we don't want to support EOL'd versions of Go in upstream cachi2.

brunoapimentel reviewed (commented):
The only major comment is about the GOTOOLCHAIN environment variable that we're setting up for `.build-config.json`. Besides that, it LGTM!

taylormadore reviewed (commented):
Overall, this LGTM aside from the output GOTOOLCHAIN=local :rocket: 

All of the local testing I've done looks good :+1:

taylormadore reviewed (approved):
Actually, I'm going to approve it anyway given that I won't be here tomorrow and that is a quick fix

eskultety commented on cachi2/core/package_managers/gomod.py:
```diff
@@ -548,10 +551,11 @@ def fetch_gomod_source(request: Request) -> RequestOutput:
+        "GOPATH": "${output_dir}/deps/gomod",
+        "GOMODCACHE": "${output_dir}/deps/gomod/pkg/mod",
+        "GOPROXY": "file://${GOMODCACHE}/cache/download",
+        "GOTOOLCHAIN": "local",
```
Thank you for spotting this, you're right. When I did this I focused solely on enforcing a consistent "base" behaviour of the target image which can default either to `GOTOOLCHAIN=local` or `GOTOOLCHAIN=auto` depending on how it was packaged on that distro and I didn't like that because it can lead to negatively surprising results. However, I then neglected the technical aspect of how the hermetic builds work and falsely assumed the user would override our value to their liking just fine, but it is the other way around actually, whatever users configure will be overriden once they do a "source" on our env file. I'll drop this.

eskultety commented on README.md:
```diff
@@ -341,8 +341,17 @@ See [docs/gomod.md](docs/gomod.md) for more details.
   [`toolchain`](https://go.dev/ref/mod#go-mod-file-toolchain) to the `go.mod` file. What this all
-  means in practice for end users is that you may not be able to process your `go.mod` file with an older version
-  of Go (and hence older cachi2) as you could in the past for various reasons.
+  means in practice for end users is that you may not be able to process your `go.mod` file with an
+  older version of Go (and hence older cachi2) as you could in the past for various reasons.
+  Many projects bump their required Go toolchain's micro release as soon as it becomes available
+  upstream (i.e. not waiting for distributions to bundle them properly). This caused problems for
+  *cachi2-v0.5.0* because the container image's version simply may not have been high enough to
+  process a given project's `go.mod` file. Therefore, *cachi2-v0.6.0* implemented a mechanism to
+  always rely on the origin 0th release of a toolchain (e.g. 1.21.0) and use the `GOTOOLCHAIN=auto`
+  setting to instruct Go to fetch any toolchain as specified by the `go.mod` file automatically,
+  hence allowing us to keep up with frequent micro version bumps. **Note that such a language
+  version would still need to be officially marked as supported by cachi2, i.e. we'd not allow Go
+  to fetch e.g. a 1.22 toolchain if the maximum supported Go version by cachi2 were 1.21!**
```
> which means [this](https://github.com/containerbuildsystem/cachi2/blob/278c0b241ad2aae69bd18be57409de80e7a7ea03/README.md#user-content-fn-go-compat-8d98370aad2c1c2101383579142caa1f) footnote is now incorrect

Technically, it was incorrect all along because any new keywords from releases newer than cachi2's base image would have still failed, I believe that's why it uses the 'there's a good chance' wording. I think it's appropriate to re-word this footnote to accurately reflect the reality of 1.21+.

> If you feel like this is important to be documented, why don't we move this to the docs/gomod.md specific docs?

I believe it is because it has a noticeable user impact, or at least it may have had based on the failures they'd get from requiring latest toolchain micro releases. Okay, I'll try to move it to the gomod docs and add a reference to that in the main README file.

eskultety replied:
Changed, let me know if you agree with the latest docs changes.

eskultety commented on cachi2/core/package_managers/gomod.py:
```diff
@@ -718,6 +735,69 @@ def _find_missing_gomod_files(source_path: RootedPath, subpaths: list[str]) -> l
     return invalid_gomod_files
 
 
+def _setup_go_toolchain(go_mod_file: RootedPath) -> Go:
```
I think it's worth noting it was me who cluttered it in the first place...

eskultety commented:
Since v6:
- docs changes as requested in review
- dropped explicit `GOTOOLCHAIN=local` setting in the generated env file to really let users to set the variable however they like
- moved the `GOTOOLCHAIN=auto` enablement to the correct commit

eskultety commented:
Since v7:
- drop the go backwards compatibility footnote as it dropping was incorrect
- convert `since cachi2-v0.6.0` from bold italic to just italic (typo)
